### PR TITLE
Add GitHub actions cicd pipeline

### LINKED ISSUE #314: Create CD/CI Pipeline for staging deployment
### Overview

As a developer, we should make sure that our deployment pipeline is automated so that developers can spend more time on critical issues rather than repetitive work. For this issue, we will leverage GHAs in order to create a CD/CI pipeline into our staging aws environment.

### Action Items

- [x] Create a GitHub action that automatically deploys the code from our `main` branch into our aws instance
  - [x] Make sure that our secrets are [injected in via our GitHub Secrets](https://docs.github.com/en/actions/security-guides/encrypted-secrets#using-encrypted-secrets-in-a-workflow)

### Resources/Instructions

Note: In the future we will probably move `main` to deploy into a production instance, and move staging deployment into a `staging` branch.

[Resources](https://hackforla.github.io/CivicTechJobs/resources/)

### PR DESCRIPTION
**Pull request by Enzyme3**
Fixes #314

### Changes
- Added two GH Actions workflows:
  - 1) on merges to main/dev, builds image and pushes to stage
  - 2) on release creation or when triggered manually, promotes build to prod

### CODE REVIEW

**Comment by Enzyme3** [comment]
- Relies on the presence `stage` and `prod` Github repo env, plus secrets stored in the environments
  - For each env, set the following GH secrets:
    - `DB_PASSWORD`
    - `AWS_ACCESS_KEY_ID`
    - `AWS_SECRET_ACCESS_KEY`
- Right now, the env points to a test DB, but will update it to the one provided to the ops team once credentials are sent
- also set Django's `ALLOWED_HOSTS` to '*' for now to accommodate healthchecks. Can follow one of the approaches listed [here](https://stackoverflow.com/questions/49828259/when-deploying-django-into-aws-fargate-how-do-you-add-the-local-ip-into-allowed) if wanted

File: app/config/settings.py [hunk 1]
```diff
@@ -28,8 +28,8 @@
 # SECURITY WARNING: don't run with debug turned on in production!
 DEBUG = os.environ.get("DEBUG", "True") != "False"
 
-ALLOWED_HOSTS = os.environ.get("DJANGO_ALLOWED_HOSTS", "localhost").split(" ")
-
+# ALLOWED_HOSTS = os.environ.get("DJANGO_ALLOWED_HOSTS", "localhost").split(" ")
+ALLOWED_HOSTS = ["*"]
```
Comment (Aveline-art) [inline app/config/settings.py:32]: This isn't really safe. Can we switch to using the workaround?

File: .github/workflows/build-deploy-stage.yml [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+  ECR_REPOSITORY: civictechjobs-fullstack-stage
+  ECS_SERVICE: civictechjobs-fullstack-stage
+  ECS_CLUSTER: incubator-prod
+  ECS_TASK_DEFINITION: ./aws/task-definition.json
+  ECS_FAMILY: civictechjobs-fullstack-stage
+  ECS_TASK_ROLE_ARN: incubator-prod-ecs-task-role
+  ECS_EXECUTION_ROLE_ARN: incubator-prod-ecs-task-role
+
+  DB_HOST: 34.132.106.11
+  DB_NAME: my-test-db
```
Comment (Aveline-art) [inline .github/workflows/build-deploy-stage.yml:18]: Are the credentials here yet? If not, we should merge this once the credentials are set.

File: .github/workflows/promote-to-prod.yml [hunk 3]
```diff
@@ -0,0 +1,81 @@
+name: Promote to Amazon ECS prod environment
+
+on:
```
Comment (Aveline-art) [inline .github/workflows/promote-to-prod.yml:3]: I'm not sure we're quite ready for this code yet. A lot can change from here to February/March. Can we separate this workflow into a separate PR?

**Review by Aveline-art — CHANGES_REQUESTED** [review]

**Comment by Aveline-art** [comment]
Just a few comments here and there. Thank you for working on this ^_^

File: app/config/settings.py (hunk 1 above)
Comment (Enzyme3) [reply app/config/settings.py]: updated to inject ALLOWED_HOSTS with IP of ECS health checker

File: .github/workflows/build-deploy-stage.yml (hunk 2 above)
Comment (Enzyme3) [reply .github/workflows/build-deploy-stage.yml]: updated to point to incubator DB

File: .github/workflows/promote-to-prod.yml (hunk 3 above)
Comment (Enzyme3) [reply .github/workflows/promote-to-prod.yml]: removed entire file from PR

File: app/entrypoint.sh [hunk 4]
```diff
@@ -1,4 +1,9 @@
 #!/bin/sh
+if [ "$TARGET" = "ECS" ]; then
+    ECS_ALLOWED_HOST=$(curl -s "${ECS_CONTAINER_METADATA_URI}" | jq -r .Networks[0].IPv4Addresses[0])
+    export ECS_ALLOWED_HOST
```
Comment (Aveline-art) [inline app/entrypoint.sh:4]: Pretty hacky but I like it 😎

File: app/config/settings.py [hunk 5]
```diff
@@ -29,8 +29,9 @@
 DEBUG = os.environ.get("DEBUG", "True") != "False"
 
 ALLOWED_HOSTS = os.environ.get("DJANGO_ALLOWED_HOSTS", "localhost").split(" ")
-
-
+if "ECS_ALLOWED_HOST" in os.environ:
+    ALLOWED_HOSTS.append(os.environ.get("ECS_ALLOWED_HOST"))
+print(f'ALLOWED_HOSTS: %s' % ','.join(ALLOWED_HOSTS))
```
Comment (Aveline-art) [inline app/config/settings.py:34]: You can remove the print? It does not look like this log is too important

File: .github/workflows/build-deploy-stage.yml [hunk 6]
```diff
@@ -0,0 +1,100 @@
+name: Deploy to Amazon ECS stage environment
+
+on:
+  push:
+    branches: [ "dev-branch-1", "main", "add-github-actions-cicd-pipeline" ]
```
Comment (Aveline-art) [inline .github/workflows/build-deploy-stage.yml:5]: Not sure we have a branch called `"add-github-actions-cicd-pipeline"`

**Review by Aveline-art — APPROVED** [review]
Everything looks good! Just a couple of things I see, but feel free to merge this in at any time. Thank you for working on this!

File: .github/workflows/build-deploy-stage.yml (hunk 6 above)
Comment (Enzyme3) [reply .github/workflows/build-deploy-stage.yml]: same here - removed as well
